### ISSUE TITLE
[Question] Class balancing in multiclass setting

### ISSUE THREAD

**Issue by cbecker** [Microsoft/LightGBM#306]
I have a doubt about multiclass clasification: does LightGBM weights each class inversely proportional to the number of training samples in that class for training?

This is generally very important to balance training data, especially on very skewed datasets.

Related to this, if LightGBM is not doing this automatically, can this be achieved by setting the `weights` metadata for each sample? Thanks!

**Comment by Laurae2** [comment]
I think you can't use `is_unbalance` is multiclass problems, you have two solutions from that point:

* You use `weights  for each sample but this is not an appropriate solution as you will notice you can not apply several weights to a single observation
* You train n-class One vs All classifier (train each class vs all others bucketed in the same class = binary classification to repeat per class). This is what LightGBM does internally, you will have access to `is_unbalance` parameter as you fall back to a binary classification problem.

To get predictions from the One vs All classifier, you must predict the likelihood of each class using the appropriate model each time (class 0 vs All, class 1 vs All, etc.). You will notice the sum of predictions is not equal to 1, this is a normal consequence of multiclass prediction: you can rescale them manually (or take the top 1 if you are interested in accuracy).

Perhaps @guolinke has a better solution than my idea.

**Comment by cbecker** [comment]
Thank you. My understanding is that, even if a 1-vs-all approach is used in training the trees, the overall loss being minimized can be interpreted as a global cost function, and the values in `weights` represent the weight for each value in this loss.

I would then expect that setting those weights (which are scalars as you said) would help generate a more balanced classifier. Is there something I am missing? I know it may not be perfect, but if there is a line-search step (or even an approximate one) to find the weight of each tree, then they will eventually approximate a minimizer of the overall cost function.

I will experiment with the weights. In my case the skewness is not that pronounced, I am not sure how much effect it would have.

**Comment by guolinke** [comment]
I thin the ```is_unbalance``` can be adapted into multi_class.
